src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -306,5 +309,108 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
+        /// Writes a DataFrame into a CSV.
+        /// Follows pandas API.
+        /// </summary>
+        /// <param name="dataFrame"><see cref="DataFrame"/></param>
+        /// <param name="csvStream">stream of CSV data to be write out</param>
+        /// <param name="separator">column separator</param>
+        /// <param name="header">has a header or not</param>
+        /// <param name="encoding">the character encoding. Defaults to UTF8 if not specified</param>
+        /// <param name="cultureInfo">culture info for handeling decimal values</param>
+        public static void WriteCsv(DataFrame dataFrame, Stream csvStream,
+                           char separator = ',', bool header = true,
+                           Encoding encoding = null, CultureInfo cultureInfo = null)
+        {
+            if (cultureInfo is null)
+            {
+                cultureInfo = CultureInfo.InvariantCulture;
+            }
+            else
+            {
+                if (cultureInfo.NumberFormat.NumberDecimalSeparator.Equals(",") && separator.Equals(','))
+                {
+                    throw new ArgumentException("Decimal separator cannot match the column separator");
+                }
+            }
+
+            using (StreamWriter csvFile = new StreamWriter(csvStream, encoding ?? Encoding.UTF8, bufferSize: DefaultStreamReaderBufferSize, leaveOpen: true))
+            {
+                if (dataFrame != null)
+                {
+                    var columnNames = dataFrame.Columns.GetColumnNames();
+
+                    if (header)
+                    {
+                        var headerColumns = string.Join(separator.ToString(), columnNames);
+                        csvFile.WriteLine(headerColumns);
+                    }
+
+                    var record = new StringBuilder();
+
+                    foreach (var row in dataFrame.Rows)
+                    {
+                        bool firstRow = true;
+                        foreach (var cell in row)
+                        {
+                            if (!firstRow)
+                            {
+                                record.Append(separator);
+                            }
+                            else
+                            {
+                                firstRow = false;
+                            }
+
+                            Type t = cell?.GetType();
+
+                            //TODO supporting strings, numbers and bools only
+
+                            if (t == typeof(bool))
+                            {
+                                record.AppendFormat(cultureInfo, "{0}", Convert.ToBoolean(cell));
+                                continue;
+                            }
+
+                            if (t == typeof(float) || t == typeof(double) || t == typeof(decimal))
+                            {
+                                record.AppendFormat(cultureInfo, "{0}", Convert.ToSingle(cell));
```

Comment:
I have a few concerns with this:

1. Converting all values to "single" can change the value. Especially for `decimal` types.
2. Should we be specifying a precision or "round trip" format here? That way the same value will be parsed back.
    - https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings#the-round-trip-r-format-specifier

I'm wondering if we should probably split the 3 types out and write them each separately, as appropriate.

Reply:
For (2) above, I just spoke with @tannergooding, and he mentioned that with .NET Core 3, you don't need to specifying "round trip" or a precision. Just calling `ToString()` on float, double, and decimal will produce the "shortest round-trippable string" for each type.

So I think we can get rid of the cast to `float` altogether and just call `ToString()` for these 3 types.

Reply:
Agreed.
It seems that "G17" for Double and "G9" for Single is even better. In addition, decimal support G but not R.